### PR TITLE
Host initialisation refactor to use multiprocess pool

### LINKED ISSUE #2292: Job host selection and preparation in workers or sub-processes
Job host selection and preparation (e.g. copying service files from suite to job host on 1st use) are currently done by the main process as part of job submission of each task. This can cause the suite to be unresponsive if these commands do not finish quickly. It will be desirable to separate the logic out and move them into the managed worker pool or sub-processes.

See also #2199 and #2144, #2214.

### PR DESCRIPTION
**Pull request by dvalters**
Host initialisation is currently done as part of the main loop in the `init_host()` method. Subprocess commands that were blocking (such as copying the files over to a remote host) are now sent to the multiprocessing pool to be carried out asynchronously, which should reduce wait times if remote hosts are busy or there are a lot of files to copy. 

Close #2292, see also #2390.

### CODE REVIEW

File: lib/cylc/task_job_mgr.py [hunk 1]
```diff
@@ -123,22 +148,73 @@ def check_task_jobs(self, suite, task_pool):
...
+        """
+        # Create init manager
+        host_init_commander = HostInitter(cmds)
+        # TODO: These might need to be executed sequentially in the pool..?
+        for cmd in cmds:
+            self.proc_pool.put_command(
+                SuiteProcContext(self.JOB_INIT_HOST, cmd),
+                self._init_host_command_callback,
+                [cmd, owner, host, user_at_host, r_suite_run_dir,
+                 host_init_commander])
```
Comment (dvalters) [inline lib/cylc/task_job_mgr.py]: I think potentially there is a race condition here as the init commands could be executed out of order once in the pool? (So for instance, you wouldn't want the copy files command executing before the directory structure had been created)

**Comment by dvalters** [comment]
Logic (sanity check):

Host initialisation commands should go into the multiprocess pool, so that they don't 'hang' the suite startup process while waiting for `init_host()` to complete. Normal tasks cannot start before the host initialisation commands have completed, so there must be a way for a task to know if its host is initialised yet. The HostInitter object (I will think of a better name...) keeps track of the host inititalisation processes and issues a callback when they have all completed, marking the host initialisation as complete.

**Comment by hjoliver** [comment]
JobHostPrep?

**Comment by dvalters** [comment]
Still some site specific tests to fix up

**Comment by dvalters** [comment]
@matthewrmshin I'm still getting some failed site-specific tests, I wonder if you have chance you could run the test-battery for these and we can compare notes so I can fix these?

```
cylc-cat-log/01-remote.t
job-submission/02-job-nn-remote-host.t
authentication/10-remote-suite-passphrase-cache.t
job-submission/08-activity-log-host.t
```

**Comment by matthewrmshin** [comment]
As discussed, only `./tests/restart/13-bad-job-host.t` is failing now.

**Comment by dvalters** [comment]
@matthewrmshin This test is actually failing on master as well (for me)

**Comment by matthewrmshin** [comment]
Healthy tests! We'll do some site tests tomorrow.

**Comment by dvalters** [comment]
The mystery slowdown on `13-bad-job-host` is also back to normal speed, as far as I could tell, when I ran it before leaving.

**Comment by matthewrmshin** [comment]
Wonderful

**Comment by matthewrmshin** [comment]
Site tests are passing in my environment.

**Comment by matthewrmshin** [comment]
(I'll take over this branch.)

**Comment by matthewrmshin** [comment]
Superseded by #2468.

**Pull request closed without merging**